client/src/components/DependentCompositeField/DependentCompositeField.js

```diff
@@ -0,0 +1,138 @@
+import { useSelector } from 'react-redux';
+import debounce from 'lodash/debounce';
+
+/*
+TODO:
+1. Make sure it actually works with redux 'n' all
+1. Figure out how we're gonna pass a formFieldSchema equivalent in through from jquery (refactor as appropriate)
+*/
+
+const DependentCompositeField = ({ formid, data, className = '', extraClass = '', formFieldSchema, childrenSchema: initialChildrenSchema, ...props }) => {
+  const Tag = data.tag || 'div';
+  const [childrenSchema, setChildrenSchema] = useState(initialChildrenSchema);
+  const reduxState = useSelector(storedState => storedState);
+
+  /**
+   * Handle changes to internal field values and update the schema appropriately.
+   * Using useCallback and debounce to prevent issues fetching a new schema while the user is still typing.
+   */
+  const handleChange = useCallback(debounce((value, state, schema, allChildrenNames) => {
+    // @TODO add a loading indicator!
+    // Call existing callback if there was one
+    if (typeof schema.onChange === 'function') {
+      schema.onChange(value);
+    }
+    // Get values from redux store but only for the relevant fields
+    const allValues = getFormValues(formid, getFormState)(state);
+    const onlyNecessaryValues = Object.fromEntries(Object.entries(allValues).filter(
+      ([key]) => allChildrenNames.includes(key)
+    ));
+    const body = {
+      fieldName: schema.name,
+      origValues: onlyNecessaryValues,
+      newValue: value,
+    };
+    // Add security ID for CSRF protected forms
+    const headers = { [FormConstants.CSRF_HEADER]: Config.get('SecurityID') };
+    // Fetch new schema from the updateChildrenURL endpoint
+    const getNewFieldSchemaApi = backend.createEndpointFetcher({
+      url: data.updateChildrenURL,
+      payloadFormat: 'json',
+    });
+    getNewFieldSchemaApi(body, headers)
+      .then(response => {
+        // @TODO what to do if schema or state isn't in the response?
+        const normalisedFields = formFieldSchema.normalizeFields(response.schema, response.state);
+        // @TODO We might be returning only the fields that got updated - if that's the case we need to merge this in with the existing children schema
+        setChildrenSchema(normalisedFields);
+        // Update state in redux form to reflect new values returned from API call for all fields with updated values
+        for (const fieldData of response.state.fields) {
+          if (fieldData.value !== onlyNecessaryValues[fieldData.name]) {
+            const valueAsNumber = Number(fieldData.value);
+            if (Number.isInteger(valueAsNumber)) {
+              fieldData.value = valueAsNumber;
+            }
+            props.meta.dispatch(change(formid, fieldData.name, fieldData.value));
+          }
+        }
+      })
+      .catch((reason) => {
+        // @TODO check what this looks like and probably give better user feedback
+        throw reason;
+      });
+  }, 300), []);
```

Comment:
300ms feels about right when typing in a text field - but it feels sluggish when selecting items from a dropdown field.

If you have any suggestions about how to deal with that discrepancy better please let me know.